### DISCUSSION TITLE
ordering_list IntegrityError

### DISCUSSION THREAD

**Discussion by joexner** [sqlalchemy/sqlalchemy#10775] [Usage Questions]
I get this strange error adding an element to the start of a mapped List of elements. I can't tell if this transaction *should* even work. Should this reordering work? By the end of the transaction (heck, statement), everything should be consistent again. It's trying to move the cell in slot 0 back to slot 1, and 1 to 2, and so on.

I'm using sqlite for a backend.

```
    def do_executemany(self, cursor, statement, parameters, context=None):
>       cursor.executemany(statement, parameters)
E       sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: organism_cell.organism_id, organism_cell.ordinal
E       [SQL: UPDATE organism_cell SET ordinal=? WHERE organism_cell.organism_id = ? AND organism_cell.ordinal = ?]
E       [parameters: [(1, '3dadbeefb68d41eaafbaac15c3ddf22f', 0), (2, '3dadbeefb68d41eaafbaac15c3ddf22f', 1), (3, '3dadbeefb68d41eaafbaac15c3ddf22f', 2), (4, '3dadbeefb68d41eaafbaac15c3ddf22f', 3)]]
E       (Background on this error at: https://sqlalche.me/e/20/gkpj)
```

Relevant bits

```
class Organism(EntityBase):
    __tablename__ = "organism"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True)

    cells: List[Cell] = association_proxy(
        target_collection="_cells_list", attr="cell", creator=_Organism_Cell)
...
    _cells_list: Mapped[List[_Organism_Cell]] = relationship(
        order_by="_Organism_Cell.ordinal",
        collection_class=ordering_list('ordinal'),
        cascade="all, delete-orphan",
        lazy="joined",
    )
}

class _Organism_Cell(EntityBase):
    __tablename__ = "organism_cell"

    organism_id = mapped_column(ForeignKey("organism.id"), primary_key=True)
    ordinal = mapped_column("ordinal", Integer, primary_key=True)
    cell_id = mapped_column(ForeignKey("cell.id"))

    organism: Mapped["Organism"] = relationship()
    cell: Mapped[Cell] = relationship(cascade="all", lazy="joined")

    def __init__(self, cell):
        self.cell = cell
```

**Comment by joexner** [comment]
Fun experiment: the list in the original test had 4 elements, and adding 1 failed, presumably because the shuffling would double-occupy some indices in transit. Adding 4 or more elements works fine w/o errors.

**Reply by zzzeek** [reply]
I've never used OrderingList and it is very old and not very widely used.  I would advise checking the warnings at https://docs.sqlalchemy.org/en/20/orm/extensions/orderinglist.html which include the specific use cases that are known to not work very well or at all.    Overall this structure is not that useful for cases where you have to move things around a bunch.

**Comment by CaselIT** [comment]
Hi,

Not sure about sqlite, but postgresql supports deferring constraints until the commit. This allows insertion duplicates rows if they are then updated or modified later in the same transaction, like in this case

**Reply by joexner** [reply]
Ah, that's kinda what I guess I hoped would happen. I'll check out deferring the constraints w/ sqlite.

I worked around the problem with a dirty hack for now, effectively removing the unique constraint by adding the last column of the join table to the primary key.
